In [10]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [11]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [12]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [13]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [14]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
df_market_datascaled = StandardScaler().fit_transform(df_market_data[["price_change_percentage_24h","price_change_percentage_7d","price_change_percentage_14d", "price_change_percentage_30d", "price_change_percentage_60d"]])

In [21]:
# Create a DataFrame with the scaled data

df_market_data_scaled = pd.DataFrame(df_market_datascaled, columns = ["price_change_percentage_24h","price_change_percentage_7d","price_change_percentage_14d", "price_change_percentage_30d", "price_change_percentage_60d"]
                                    )
# Copy the crypto names from the original data

df_market_data_scaled["coin_id"] = df_market_data.index

# Set the coinid column as index

df_market_data_scaled = df_market_data_scaled.set_index("coin_id")

# Display sample data
df_market_data_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  
coin_id                                    
bitcoin                         -0.067495  
ethereum                        -0.273483  
tether                           0.008005  
ripple                          -0.373164  
bitcoin-cash                    -0.291203

---

### Find the Best Value for k Using the Original Data.

In [26]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

In [29]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list


for i in k:
    k_model = KMeans(n_clusters=i, random_state=1)
    k_model.fit(df_market_data_scaled)
    inertia.append(k_model.inertia_)
    

C:\Users\clavo\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1036: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


In [30]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}


# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)

In [31]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4 

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [33]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters =4, random_state = 1)

In [34]:
# Fit the K-Means model using the scaled data
model.fit(df_market_data_scaled)

KMeans(n_clusters=4, random_state=1)

In [37]:
# Predict the clusters to group the cryptocurrencies using the scaled data
predicted_cluster = model.predict(df_market_data_scaled) 

# Print the resulting array of cluster values.
predicted_cluster

array([1, 1, 3, 3, 1, 1, 1, 1, 1, 3, 3, 3, 3, 1, 3, 1, 3, 3, 1, 3, 3, 1,
       3, 3, 3, 3, 3, 3, 1, 3, 3, 3, 0, 1, 3, 3, 2, 3, 3, 3, 3])

In [39]:
# Create a copy of the DataFrame
marketdata_predictions = df_market_data_scaled.copy()

In [41]:
# Add a new column to the DataFrame with the predicted clusters

marketdata_predictions["crypto_cluster"] = predicted_cluster

# Display sample data
marketdata_predictions.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  crypto_cluster  
coin_id                                                    
bitcoin                         -0.067495               1  
ethereum                        -0.273483               1  
tether                           0.008005               3  
ripple                          -0.373164               3  
bitcoin-cash                    -0.291203               1

In [46]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

marketdata_predictions.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d", 
    by="crypto_cluster", 
    hover_cols="coin_id") 

:NdOverlay   [crypto_cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [47]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [49]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
marketdata_pca = pca.fit_transform(df_market_data_scaled) 
# View the first five rows of the DataFrame. 
marketdata_pca[:5]

array([[ 0.79510406, -0.66290603, -0.15388459],
       [ 0.41376913, -1.04777288,  0.15762251],
       [-0.19550822,  0.51826138, -0.20153516],
       [-0.26074757,  0.34060069, -0.14574095],
       [ 1.96101883, -2.23972436, -0.18298129]])

In [50]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.47862164, 0.26608254, 0.1684978 ])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.47862164, 0.26608254, 0.1684978 

In [54]:
# Create a new DataFrame with the PCA data.
market_data_pca = pd.DataFrame(marketdata_pca,columns= ["PC1","PC2","PC3"])

# Creating a DataFrame with the PCA data


# Copy the crypto names from the original data
market_data_pca["coin_id"] = df_market_data.index

# Set the coinid column as index
market_data_pca = market_data_pca.set_index("coin_id")

# Display sample data
market_data_pca.head()

PC1       PC2       PC3
coin_id                                   
bitcoin       0.795104 -0.662906 -0.153885
ethereum      0.413769 -1.047773  0.157623
tether       -0.195508  0.518261 -0.201535
ripple       -0.260748  0.340601 -0.145741
bitcoin-cash  1.961019 -2.239724 -0.182981

---

### Find the Best Value for k Using the PCA Data

In [55]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

In [56]:
# Create an empy list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model = KMeans(n_clusters=i, random_state=1)
    k_model.fit(market_data_pca)
    inertia.append(k_model.inertia_)


C:\Users\clavo\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1036: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


In [57]:
# Create a dictionary with the data to plot the Elbow curve
elbowpca_data = {"k": k, "inertia": inertia}
# Create a DataFrame with the data to plot the Elbow curve
df_elbowpca = pd.DataFrame(elbowpca_data)

In [58]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
df_elbowpca.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** no
 

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [59]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters =4, random_state = 1)

In [62]:
# Fit the K-Means model using the PCA data
model.fit(market_data_pca)

KMeans(n_clusters=4, random_state=1)

In [74]:
# Predict the clusters to group the cryptocurrencies using the PCA data
market_clusters = model.predict(market_data_pca)
# Print the resulting array of cluster values.
print(market_clusters)

[1 1 3 3 1 1 1 1 1 3 3 3 3 1 3 1 3 3 1 3 3 1 3 3 3 3 3 3 1 3 3 3 0 1 3 3 2
 3 3 3 3]


In [75]:
# Create a copy of the DataFrame with the PCA data
market_pca_predict = market_data_pca.copy()

# Add a new column to the DataFrame with the predicted clusters

market_pca_predict["marketclusters"]= market_clusters

# Display sample data
market_pca_predict

PC1       PC2       PC3  marketclusters
coin_id                                                           
bitcoin               0.795104 -0.662906 -0.153885               1
ethereum              0.413769 -1.047773  0.157623               1
tether               -0.195508  0.518261 -0.201535               3
ripple               -0.260748  0.340601 -0.145741               3
bitcoin-cash          1.961019 -2.239724 -0.182981               1
binancecoin           1.338761 -1.017717 -0.222509               1
chainlink             0.465465 -2.840707  0.745956               1
cardano               0.690398 -1.494549  0.399413               1
litecoin              0.391531 -0.701588  0.176794               1
bitcoin-cash-sv      -0.258488 -0.124559 -0.581406               3
crypto-com-chain     -1.427382  1.084171 -1.051990               3
usd-coin             -0.203583  0.532440 -0.221013               3
eos                  -0.533999  0.241185 -0.432517               3
monero                2.943633 -1.206637  0.937170               1
tron                 -0.149095 -0.117440 -0.119299               3
tezos                -0.541640 -1.212586  0.224646               1
okb                  -1.245232  1.284823  0.312635               3
stellar              -0.777384 -0.041335  0.024432               3
cosmos                0.684804 -1.559531  0.835680               1
cdai                 -0.179774  0.463768 -0.348912               3
neo                  -1.040693  0.453352 -0.672750               3
wrapped-bitcoin       0.778979 -0.646726 -0.170150               1
leo-token            -0.710825  0.872126 -0.502457               3
huobi-token          -0.240953  0.258779 -0.076616               3
nem                   0.616022  1.430832  0.257337               3
binance-usd          -0.180509  0.507697 -0.237092               3
iota                 -0.597839 -0.524172 -0.396794               3
vechain              -2.045414  0.649799 -1.478250               3
zcash                 0.467782 -1.028666  0.476763               1
theta-token           0.148114  2.962392  1.163818               3
dash                 -0.516422 -0.399828  0.219206               3
ethereum-classic     -0.396475 -0.028171 -0.092574               3
ethlend              -2.876399  1.046329  4.238900               0
maker                 0.135667 -0.399542  0.204228               1
havven               -2.227749  0.118763  0.764482               3
omisego               0.333054  0.573802 -1.922353               3
celsius-degree-token  7.110555  3.206833  0.514679               2
ontology             -2.176115  0.421112 -0.337958               3
ftx-token            -0.017398 -0.431146 -0.313553               3
true-usd             -0.165950  0.490402 -0.246053               3
digibyte             -0.309084  0.267834 -1.545373               3

In [77]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.


market_pca_predict.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="marketclusters"
)

DataError: Supplied data does not contain specified dimensions, the following dimensions were not found: ['price_change_percentage_24h', 'price_change_percentage_7d']

PandasInterface expects tabular data, for more information on supported datatypes see http://holoviews.org/user_guide/Tabular_Datasets.html

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [30]:
# Composite plot to contrast the Elbow curves

In [31]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** 